### PR TITLE
Move the CompactSize, Vector, Array and Option encodings to a new zcash_encoding crate.

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

File: components/zcash_encoding/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "zcash_encoding"
+description = "Binary encodings of types used througout the Zcash ecosystem."
```
Comment (daira) [inline components/zcash_encoding/Cargo.toml:3]:
```suggestion
description = "Binary encodings of types used throughout the Zcash ecosystem."
```

**Review by daira — APPROVED** [review]
utACK modulo remaining clippy lints.

File: components/zcash_encoding/README.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# zcash_encodings
+
+Zcash-related types that require stable binary encodings for use throughout the 
+Zcash ecosystem.
```
Comment (str4d) [inline components/zcash_encoding/README.md:4]:
```suggestion
This library provides common encoding and decoding operations for stable binary
encodings used throughout the Zcash ecosystem.
```

**Review by str4d — CHANGES_REQUESTED** [review]
LGTM overall.

File: components/zcash_encoding/src/lib.rs [hunk 3]
```diff
@@ -1,12 +1,25 @@
...
+#![deny(unsafe_code)]
+
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
 use nonempty::NonEmpty;
 use std::io::{self, Read, Write};
 
 const MAX_SIZE: usize = 0x02000000;
 
+/// Namespace for functions for compact encoding of integers in the range
+/// 0x0..=0x02000000
```
Comment (str4d) [inline components/zcash_encoding/src/lib.rs:18]:
```suggestion
/// Namespace for functions for compact encoding of integers.
///
/// This codec requires integers to be in the range `0x0..=0x02000000`, for compatibility
/// with Zcash consensus rules.
```

File: components/zcash_encoding/src/lib.rs [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
 pub struct CompactSize;
 
 impl CompactSize {
+    /// Read an integer encoded in compact form.
```
Comment (str4d) [inline components/zcash_encoding/src/lib.rs:22]:
```suggestion
    /// Reads an integer encoded in compact form.
```

File: components/zcash_encoding/src/lib.rs [hunk 5]
```diff
@@ -63,9 +77,14 @@ impl CompactSize {
     }
 }
 
+/// Namespace for functions that perform encoding of vectors that represent a vector as a
+/// CompactSize-encoded integer specifying the length of the vector, followed by the encoding of
+/// each element of the vector.
```
Comment (str4d) [inline components/zcash_encoding/src/lib.rs:82]:
```suggestion
/// Namespace for functions that perform encoding of vectors.
///
/// The length of a vector is restricted to at most `0x02000000`, for compatibility with
/// the Zcash consensus rules.
```

File: components/zcash_encoding/src/lib.rs [hunk 6]
```diff
@@ -74,6 +93,9 @@ impl Vector {
         Array::read(reader, count, func)
     }
 
+    /// Writes a slice of values by writing CompactSize-encoded integer specifying the length of
```
Comment (str4d) [inline components/zcash_encoding/src/lib.rs:96]:
```suggestion
    /// Writes a slice of values by writing [`CompactSize`]-encoded integer specifying the length of
```

File: components/zcash_encoding/src/lib.rs [hunk 7]
```diff
@@ -95,16 +119,23 @@ impl Vector {
     }
 }
 
+/// Namespace for functions that perform encoding of array contents.  This is similar to the
+/// [`Vector`] encoding except that no length information is written as part of the encoding, so
+/// length must be statically known or obtained from other parts of the input stream.
```
Comment (str4d) [inline components/zcash_encoding/src/lib.rs:124]:
```suggestion
/// Namespace for functions that perform encoding of array contents.
///
/// This is similar to the [`Vector`] encoding except that no length information is
/// written as part of the encoding, so length must be statically known or obtained from
/// other parts of the input stream.
```

File: zcash_primitives/src/transaction/components/tze.rs [hunk 8]
```diff
@@ -2,15 +2,14 @@
 #![cfg(feature = "zfuture")]
 
 use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
-
```
Comment (str4d) [inline zcash_primitives/src/transaction/components/tze.rs:5]: Slight misunderstanding of my previous comment: by our existing convention, `use byteorder` would be grouped with `use zcash_encoding`.

**Pull request merged**